extensions/context/src/main/java/io/opentelemetry/context/Context.java

```diff
@@ -0,0 +1,302 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Copyright 2015 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.opentelemetry.context;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.Nullable;
+
+/**
+ * A context propagation mechanism which can carry scoped-values across API boundaries and between
+ * threads.
+ *
+ * <p>A Context object can be {@link #attach attached} to the {@link ContextStorage}, which
+ * effectively forms a <b>scope</b> for the context. The scope is bound to the current thread.
+ * Within a scope, its Context is accessible even across API boundaries, through {@link #current}.
+ * The scope is later exited by {@link Scope#close()} closing} the scope.
+ *
+ * <p>Context objects are immutable and inherit state from their parent. To add or overwrite the
+ * current state a new context object must be created and then attached, replacing the previously
+ * bound context. For example:
+ *
+ * <pre>
+ *   Context withCredential = Context.current().withValue(CRED_KEY, cred);
+ *   withCredential.wrap(new Runnable() {
+ *     public void run() {
+ *        readUserRecords(userId, CRED_KEY.get());
+ *     }
+ *   }).run();
+ * </pre>
+ *
+ *
+ *
+ * <p>Notes and cautions on use:
+ *
+ * <ul>
+ *   <li>Every {@code attach()} must be followed by a {@code Scope#close()}. Breaking these rules
+ *       may lead to memory leaks.
+ *   <li>While Context objects are immutable they do not place such a restriction on the state they
+ *       store.
+ *   <li>Context is not intended for passing optional parameters to an API and developers should
+ *       take care to avoid excessive dependence on context when designing an API.
+ * </ul>
+ */
+public final class Context {
+
+  /** Returns the root {@link Context} which all other {@link Context} are derived from. */
+  public static Context root() {
```

Comment:
Nice work.

What is the motivation for exposing the root?  When is this useful?

Reply:
Same comment. I'd be happier if we don't expose it *now*, in order to keep things simple.

Reply:
Any reason to call it root?  Perhaps `empty` makes more sense?

Reply:
Sorry I missed this comment - @tylerbenson I'm not sure about `empty` because a context override could change the default context to include something (needed if someone needs to stuff mutable state in there). I could rename to `defaultContext()` (not `default()` because Java), but any thoughts on which sounds best?